File: Algorithms/tree/938.range-sum-of-bst/range-sum-of-bst.go
// Package _38_range_sum_of_bst - package brief introduce
/* Copyright 2019 All Rights Reserved. */
/* range-sum-of-bst - file brief introduce */
/*
modification history
-----------------------------
2020/2/4 1:50 PM, by gosyang, create
*/
/*
DESCRIPTION
给定二叉搜索树的根结点 root，返回 L 和 R（含）之间的所有结点的值的和。

二叉搜索树保证具有唯一的值。
提示：

树中的结点数量最多为 10000 个。
最终的答案保证小于 2^31。
*/
package _38_range_sum_of_bst


// Definition for a binary tree node.
type TreeNode struct {
	Val int
    Left *TreeNode
    Right *TreeNode
}

func rangeSumBST(root *TreeNode, L int, R int) int {
	if root == nil {
		return 0
	}
	if root.Val > R {
		return rangeSumBST(root.Left, L, R)
	}
	if root.Val < L {
		return rangeSumBST(root.Right, L, R)
	}
	return rangeSumBST(root.Left, L, R) + root.Val + rangeSumBST(root.Right, L, R)
}

File: Algorithms/tree/938.range-sum-of-bst/range-sum-of-bst.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
########################################################################
# 
# Copyright (c) 2017 gosyang. All Rights Reserved
# 
########################################################################
 
"""
File: range-sum-of-bst.py
Author: gosyang
Date: 2020/02/04 16:06:38
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque

class Solution(object):
    def rangeSumBST(self, root, L, R):
        """
        :type root: TreeNode
        :type L: int
        :type R: int
        :rtype: int
        """
        res = 0
        deq = deque([root],)
        while deq:
            # root = deq.popleft()
            root = deq.pop()
            if not root:
                continue
            if root.val > R:
                deq.append(root.left)
                continue
            if root.val < L:
                deq.append(root.right)
                continue
            res += root.val
            deq.append(root.left)
            deq.append(root.right)

        return res
        
        # 非递归的话，用deq模拟栈，popleft就是宽搜了,也可以，差不多 